### PR TITLE
machines: Port VMs list to ListingTable component

### PR DESCRIPTION
**Pull request by KKoukiou**
- [x] blocked on #13562

### CODE REVIEW

**Comment by garrett** [comment]
We still have the issue with PatternFly 4 adapting the lists too early.

When PF4 has a breakpoint of 768px width, it's actually 1108 for us, as the sidebar is 340px. (This does use the intended content area, but there's a mismatch between content and browser width as we use iframes.)

That is, PF4 assumes a sidebar is in the same page when calculating breakpoints for the table _and_ the sidebar.

I'm wondering if this is a dealbreaker on switching to PF4 lists (for the time being), as it impacts desktop usage and makes the lists hard to follow.

I've looked through a bunch of documentation, code, GitHub issues, etc. and haven't found a way around this. Perhaps I'm missing something, but I think PF4 is just made for people using sidebars _without_ iframes.

**Comment by garrett** [comment]
So we could probably do a bit of subtraction of the standard breakpoints and subtract our sidebar from everything until the sidebar goes away*.

https://github.com/patternfly/patternfly-next/blob/master/src/patternfly/sass-utilities/scss-variables.scss#L156-L162

(* The sidebar disappears in mobile mode, so we wouldn't need to subtract at those small sizes.)

**Comment by garrett** [comment]
Note: CSS variables aren't supported in media queries by any browser. (Yet?) It's planned, but not implemented (afaict).

However, PF4 does have mappings from SASS to CSS variables for the breakpoints @ https://github.com/patternfly/patternfly-next/blob/master/src/patternfly/_variables.scss#L185-L191 — but I guess these are for reference outside of media queries and/or future proofing (perhaps).

Meanwhile, PF4's table widget uses SASS breakpoints @ https://github.com/patternfly/patternfly-next/blob/master/src/patternfly/components/Table/table.scss#L27.

**Comment by KKoukiou** [comment]
> Note: CSS variables aren't supported in media queries by any browser. (Yet?) It's planned, but not implemented (afaict).
> 
> However, PF4 does have mappings from SASS to CSS variables for the breakpoints @ https://github.com/patternfly/patternfly-next/blob/master/src/patternfly/_variables.scss#L185-L191 — but I guess these are for reference outside of media queries and/or future proofing (perhaps).
> 
> Meanwhile, PF4's table widget uses SASS breakpoints @ https://github.com/patternfly/patternfly-next/blob/master/src/patternfly/components/Table/table.scss#L27.

From the code it seems that the only way to affect the breakpoints (max-width) in the media queries to to change the breakpoint variables in scss, and recompile the patternfly. @martinpitt to you maybe see some other solution here?
@garrett changing the source code and recompiling patternfly doesn't seam very appealing to me tbh.

**Comment by garrett** [comment]
I've been digging into what we should do with regard to breakpoints.

In doing so, I've been referring to a recent article "[What browser sizes do people use the most?](https://uxplanet.org/what-browser-sizes-do-people-use-the-most-2553d4f71fc8)"

## PF4's breakpoints, with naïve sidebar subtraction

  | PF4 | _without sidebar & dashboards_ | without sidebar only
-- | -- | -- | --
xs | 0 | 0 | 0
sm | 576 | 236 | 336
**md** | 768 | 428 | 528
lg | 992 | 652 | 752
xl | 1200 | 860 | 960
2xl | 1450 | 1110 | 1210



## Various pixel width values (and what they mean)

Pixel value | Association
--: | :--
< 768 | shell switches to mobile
340 | current navbar, with dashboard switcher
240 | host navbar, without dashboard switcher
100 | dashboard switcher
**428** | minimum content width when page is 768 (nav + dash)
528 | minimum content width when page is 768 (just nav)
410 | pixel value width of large phones (iPhone 7/8 Plus)
419 | max pixel value width according to recent browser size article’s stats

## Conclusions?

Perhaps we should consider "420" as a magic number. Widths below are mobile, widths above are desktop — even with our sidebar. 

Precisely 428px could also work, according to the simplistic approach of just subtracting the sidebar. Mobile devices would still get deconstructed lists and desktops would get lists that don't fall apart.

My recommendation is that we redefine the breakpoints by subtracting the pixel value of the sidebar. We should probably also use `!default` in the CSS in our redefined variables, so we could override our overrides for the shell, if we do wind up basing it on PF4 at some point in the future.

**Comment by garrett** [comment]
Where's the right place to put PF4 variable overrides?

I've tried 
- pkg/lib/variables.scss
- pkg/lib/_global-variables.scss
- pkg/lib/page.scss

...And none work.

Where should it go?

**Comment by garrett** [comment]
Since the list is currently a mess on mobile, I considered what we could do to port to PF4 and also have mobile working.

1. Make sure the children do not overflow in mobile (and consider whether or not it should overflow in desktop mode too)
2. There's no additional indentation horizontally, as we're already space-constrained as-is. Therefore, we need to rely on a bunch of visual clues. (Spacing, alignment, lighter weight separators, font weight, blue highlight on the side, etc.)
3. Allow the tab tertiary nav to overflow and center it to make it obvious that it's navigation (offsetting it from the rest)
4. Move "Add" actions to the left and sorting to the right (or after, but in this case it isn't needed), as the sorting directly affects the list.
5. Use rich cards for the disk sub-list in this case (in other words not a nested table)

The mobile mode is based on the width of an iPhone SE. In a majority of cases, phones will be a little wider. I targeted the extreme case to make sure the layout would work.

While this is a mockup for disks, we should consider a very similar layout for the other tabs.

vm-disks

**Comment by garrett** [comment]
Here's what tertiary nav looks like at top-level (without being nested in anything):

Screen Shot 2020-03-24 at 17 37 44

It's really an extremely bad pattern and we should avoid it and wrap like my design above.

**Comment by KKoukiou** [comment]
@garrett the VM list is being adjusted according to the mockup - the Disks list (same for NICs) need some discussion and I suggest that they should be moved to another PR, since this does really affect the nested lists somehow.
So, please take a look in the VM list state and let's move the Disks and Network lists into a seperate PR if possible.
Here are some issues with the nested lists:
* There is no such thing as a disk name as shown in the mockup (you can take the file name but it's basically extra information that can be extracted from the source)
* There is not always information about the usage data of the list, so progress bar might not be possible always
* When available we show extra disk data, like io, discard policy, cache info and so on. These really need labels and I am not sure how to display them in the current mockup.

**Comment by garrett** [comment]
>     There is no such thing as a disk name as shown in the mockup (you can take the file name but it's basically extra information that can be extracted from the source)

Yep, it's the basename of the filename. It's a unique-ish identifier. It's unique if it's in the same directory as all the others.

>    There is not always information about the usage data of the list, so progress bar might not be possible always

Sure, it can be dropped when it isn't available.

>    When available we show extra disk data, like io, discard policy, cache info and so on. These really need labels and I am not sure how to display them in the current mockup.

Can you provide a screenshot of that?

All of this could be answered on the other PR (when it exists) too.

**Comment by KKoukiou** [comment]
@garrett I just realized that this need little bit attention from you.
* the toggles are misalligned.
* you also mentioned that you probably want to seperate the CSS rules for the new tables from the ones for the old.

**Comment by garrett** [comment]
> @garrett I just realized that this need little bit attention from you.

Yep. I think we mentioned those two items in IRC last week. Thanks for following up here. :+1: 

I have the start of a patch for this, but it'll take a little more work still. I've been prioritizing other things, but will switch back to this PR this week (probably starting today, in a little bit).

**Comment by garrett** [comment]
Just to give an update on my status with this:

In the midst of other PRs, I've been working on a local fork of this branch @ https://github.com/KKoukiou/cockpit/compare/patternfly-4-tables-machines...garrett:pf4-tables-WIP?expand=1 with https://patternfly-next-pr-2757.surge.sh/documentation/core/components/tabs as a reference (to adapt the tertiary nav to look like the upcoming tab component).

My branch is a WIP, but I'll get it to a state soon where I'll push commits here. It's a pretty big overhaul and I'm not quite done enough yet. Hopefully soon. I think it's getting there now.

A little while ago (a week or two or three ago? time is relative?) I changed focus after some of our discussions on IRC to try to keep it as pure PF4 as possible (with some overrides and adaptations to make it all work together) and ditch the old listing CSS for the new tables. (We'll keep the old CSS around as long as other pages are using it for their listing). It's basically sidestepping it with new class names and new files, so we can eventually deprecate and drop the old code (as there's a _lot_ of it and some would never be used and others might get in the way).

Anyway, as it's a WIP on a WIP, it still needs a bit of work before I merge it back here. I'll keep juggling this with the other PRs I'm working on, to move it (and the others forward).

**Comment by garrett** [comment]
I reworked, squashed, resorted, etc. my WIP changes and shared them here.

It looks mostly right for the Machines page now, but I broke the Firewall lists, as we're not depending on the old listing.scss file now. Therefore, it needs some porting.

So, next week, I will fix things by doing:

- [ ] Fix the position of the expander icon (">" at the start of the lists)
- [ ] Port over the necessary styling for the standard lists, as used in the Firewall page

Meanwhile, I think there's a lot of progress and we're getting closer?

File: pkg/lib/listing.scss [hunk 1]
```diff
@@ -631,6 +617,25 @@ tbody tr.listing-ct-noexpand {
     cursor: default;
 }
 
+/* Fix up nav-tabs-pf to work properly */
```
Comment (KKoukiou) [inline pkg/lib/listing.scss]: Is this something that you would like to report back to PF?

Comment (garrett) [reply pkg/lib/listing.scss]: I don't know where that comes from. It's in listing.scss, which is the old listing code. And the `nav-tabs-pf` name means it's PF4. And the fact that there's `0px` and a lot of `!important` means the CSS comes from someone else, not me, so I don't know.

I reverted the changes in listing from master, to undo the intertwined changes from this PR and the pre-existing listing widgets. We need a clean break, hence new CSS for the JS+HTML in the code (and this is old CSS).

**Comment by garrett** [comment]
I've fixed the issues with the lists used on other pages.

I think all that's left is to fix the ">" expander and go over the code again, to (possibly) clean it up a little and add additional comments? I'll work on that tomorrow.

**Comment by garrett** [comment]
I've improved better mobile support and added a few additional bugfixes for both desktop and mobile.

Next up: Fixing the ">" expander icon.

File: pkg/lib/cockpit-components-table.scss [hunk 2]
```diff
@@ -1,14 +1,90 @@
...
+        > .pf-c-button {
+            padding-top: var(--pf-global--spacer--sm);
+            padding-bottom: var(--pf-global--spacer--sm);
+            vertical-align: top;
+            // Not every browser supports text-top, so define top (above) as a fallback
+            vertical-align: text-top;
+        }
+    }
+
+    // Shrink buttons down a little
```
Comment (KKoukiou) [inline pkg/lib/cockpit-components-table.scss:70]: Why do we want to shrink the buttons  here?

File: pkg/lib/cockpit-components-table.scss [hunk 3]
```diff
@@ -1,14 +1,90 @@
...
+    }
+
+    &-empty {
+        tr td {
+            // FIXME: Not sure if this is needed. Was a direct port.
+            font-weight: normal;
+        }
+    }
+
+    // Fix toggle button alignment
```
Comment (KKoukiou) [inline pkg/lib/cockpit-components-table.scss:53]: The toggle button upstream behaves correctly. Can you confirm or reject?

File: pkg/lib/cockpit-components-table.scss [hunk 4]
```diff
@@ -1,14 +1,90 @@
...
-// Display the actions next to the header inline
-header {
-    display: flex;
-    > :only-child {
+.ct-table {
+    &.pf-m-compact {
+        > thead, > tbody {
+            > tr:not(.pf-c-table__expandable-row) {
+                // We actually want the normal font size for our lists
+                --pf-c-table-cell--FontSize: var(--pf-global--FontSize--md);
```
Comment (KKoukiou) [inline pkg/lib/cockpit-components-table.scss:9]: PF has smaller font for compact lists.
I think this makes sense or if not it should be discussed also with PF team.
 
In this case:
--pf-c-table-cell--FontSize: var(--pf-c-table--m-compact--FontSize);
--pf-c-table--m-compact--FontSize: var(--pf-global--FontSize--sm);

File: pkg/lib/cockpit-components-table.scss (hunk 2 above)
Comment (garrett) [reply pkg/lib/cockpit-components-table.scss:70]: Alignment is completely whacked with the normal sized buttons.

I can try to accommodate it though. I don't remember if this was before or after I undid the font-size shrinking due to `pf-m-compact`.

File: pkg/lib/cockpit-components-table.scss (hunk 3 above)
Comment (garrett) [reply pkg/lib/cockpit-components-table.scss:53]: But upstream alignment actually _doesn't_ work when there are widgets in a row, especially buttons. Hence the need for this.

File: pkg/lib/cockpit-components-table.scss (hunk 4 above)
Comment (garrett) [reply pkg/lib/cockpit-components-table.scss:9]: We probably should remove `pf-m-compact` from our JSX/HTML and redo the padding ourselves. That's what we really want from the pf-m-compact in this way. Then we wouldn't have to change the font back... just the padding. We're already having to do alignment and padding changes anyway, so... it probably makes sense?

File: pkg/lib/cockpit-components-listing-panel.jsx [hunk 5]
```diff
@@ -0,0 +1,161 @@
...
+            if ('presence' in props.tabRenderers[tabIdx])
+                tabPresence = props.tabRenderers[tabIdx].presence;
+            else
+                tabPresence = 'default';
+            // the active tab is covered by separate logic
+            if (tabPresence == 'always')
+                loadedTabs[tabIdx] = true;
+        }
+        // ensure the active tab is loaded
+        loadedTabs[props.initiallyActiveTab || 0] = true;
```
Comment (skobyda) [inline pkg/lib/cockpit-components-listing-panel.jsx:53]: Here you declare loadedTabs as object but later access it as based on index? Shouldn't loadedTabs be array?

File: pkg/lib/cockpit-components-listing-panel.jsx [hunk 6]
```diff
@@ -0,0 +1,160 @@
...
```
Comment (skobyda) [inline pkg/lib/cockpit-components-listing-panel.jsx:53]: Here you declare loadedTabs as object but later access it as based on index? Shouldn't loadedTabs be array?

File: pkg/machines/components/vm/vmActions.jsx [hunk 7]
```diff
@@ -17,110 +17,164 @@
...
+                              id={`${id}-delete`}
+                              variant='danger'
+                              tooltip={<Tooltip id={`${id}-delete-tooltip`}>
+                                  {_("This VM is transient. Shut it down if you wish to delete it.")}
+                              </Tooltip>}
+                              isDisabled>
+                    {_("Delete")}
+                </DropdownItem>
+            );
+        }
```
Comment (skobyda) [inline pkg/machines/components/vm/vmActions.jsx:151]: This needs to be restyled. Like this, first delete button is pushed into array and rendered always, and if vm is persistend, then second Delete link is also rendered. 
Checkout the [screenshot](https://logs.cockpit-project.org/logs/pull-13474-20200426-185359-a3c0d528-centos-8-stream/TestMachinesDBus-testDelete-centos-8-stream-127.0.0.2-2201-FAIL.png) of this situation.
I think correct solution should be something like:
```
        if (vm.persistent) {
            dropdownItems.push(
￼                 <DropdownItem className='pf-m-danger' key={`${id}-delete`} id={`${id}-delete`} onClick={() => toggleDeleteModal(true)}>
￼                     {_("Delete")}
￼                 </DropdownItem>
￼             );
￼         } else {
￼             dropdownItems.push(
￼                 <DropdownItem key={`${id}-delete`}
￼                               id={`${id}-delete`}
￼                               variant='danger'
￼                               tooltip={<Tooltip id={`${id}-delete-tooltip`}>
￼                                   {_("This VM is transient. Shut it down if you wish to delete it.")}
￼                               </Tooltip>}
￼                               isDisabled>
￼                     {_("Delete")}
￼                 </DropdownItem>
￼             );
￼         }
```

Comment (garrett) [reply pkg/machines/components/vm/vmActions.jsx:151]: We also need to adjust the order.

The non-maskable interrupt should be right before delete. I think the order we want would be:

> - Pause
> --
> - Shutdown
> - Force shut down
> --
> - Restart
> - Force restart
> --
> - Misc. stuff here
> --
> - Delete

So the order should be:

1. Common actions
2. Miscellaneous, rarely used
3. _Danger, [Will Robinson](https://cultural-phenomenons.fandom.com/wiki/Danger,_Will_Robinson)!_

File: pkg/lib/cockpit-components-listing-panel.jsx (hunk 6 above)
Comment (KKoukiou) [reply pkg/lib/cockpit-components-listing-panel.jsx:53]: That is not new code - it's just code move from pkg/lib/cockpit-components-listing.jsx

It's just an object indexed by numbers, which indeed is confusing.
Let's fix it in another PR, since this PR did not introduce this.

File: pkg/lib/listing.scss [hunk 8]
```diff
@@ -85,7 +85,6 @@
 tbody > tr.listing-ct-item {
     border: 1px solid transparent;
     border-top: 1px solid var(--color-ct-list-border);
-    border-bottom: 1px solid var(--color-ct-list-border);
```
Comment (KKoukiou) [inline pkg/lib/listing.scss:88]: I guess you want to remove both border-top and border-bottom lines here?

Comment (garrett) [reply pkg/lib/listing.scss:88]: We do want border lines between the cells, but I'll check.

**Comment by KKoukiou** [comment]
@garrett this is done from my side - there are two items more:

* The rule for making the test of empty list be centered is overwritten from a PF4 rule - so it's appearing to the right.

* Secondly as we discussed on IRC - the animation when expanding should have linked a issue to PF4 repository. I don't want to merge PF4 overwrites without link to open issue.

**Comment by garrett** [comment]
>     The rule for making the test of empty list be centered is overwritten from a PF4 rule - so it's appearing to the right.

Looking into this. IIRC: the issue is related to the buttons in the list and how that throws PF4 out of whack, as they make (wrong) assumptions about how things should be aligned, and PF4 falls apart when heights differ (as any embedded widget, such as a button) breaks that assumption.

>    Secondly as we discussed on IRC - the animation when expanding should have linked a issue to PF4 repository. I don't want to merge PF4 overwrites without link to open issue.

I'm refactoring it locally to make it better to push upstream. Meanwhile, I have a tab open with the issue creation and the old (current in this PR), that I'm planning to replace with the revised version.

**Comment by garrett** [comment]
I've modified the animation CSS to be more generic, affecting all the contents (instead of targeting the Cockpit-specific tab row and body), so it can be upstreamable. I'm also using PF4 timings and easings too.

**Comment by garrett** [comment]
I've opened an upstream issue about the table expand animation at https://github.com/patternfly/patternfly/issues/3053 and have linked to it as a comment in our code.

**Comment by garrett** [comment]
The animation code has been tested as working in Firefox and Chrome.

However, it doesn't animate in WebKit (Epiphany), and causes a little weird jump.

I guess I'll look into the WebKit issue Monday. (Should we make a fixup PR later? Should an off-by-a-pixel jump in WebKit to hold back the PR?)

Anyway, it's the weekend for all of us, so it will hopefully just be a minor delay on Monday regardless. This PR isn't landing today anyway.

**Comment by KKoukiou** [comment]
Selenium tests are broken - I will try to replace them with verify instead of fixing and delete them.

File: pkg/lib/cockpit-components-listing-panel.scss [hunk 9]
```diff
@@ -36,6 +39,7 @@
         }
       }
     }
+    }
```
Comment (martinpitt) [inline pkg/lib/cockpit-components-listing-panel.scss]: That looks like a wrong indentation somewhere?

File: pkg/lib/cockpit-components-table.jsx [hunk 10]
```diff
@@ -242,7 +243,6 @@ ListingTable.defaultProps = {
     columns: [],
     rows: [],
     actions: [],
-    className: 'ct-table',
```
Comment (martinpitt) [inline pkg/lib/cockpit-components-table.jsx]: That change deserves an explanation, and a separate commit. It's now hardcoded above in render(), but it's not clear what happens here. (Also, using `props` as a variable is confusing by itself, as one easily misreads it as `this.props`)

File: pkg/packagekit/history.jsx [hunk 11]
```diff
@@ -106,6 +106,7 @@ export class History extends React.Component {
 
         return (
             <ListingTable caption={_("Update History")}
+                          aria-label={_("Updates History")}
```
Comment (martinpitt) [inline pkg/packagekit/history.jsx:109]: Please use the same string. Also, perhaps ListingTable could just default to setting aria-label to caption if not specified? That should fix more cases in a more generic fashion.

File: pkg/packagekit/updates.jsx [hunk 12]
```diff
@@ -369,6 +369,7 @@ function UpdatesList(props) {
 
     return (
         <ListingTable className="available"
+                aria-label={_("Available Updates")}
```
Comment (martinpitt) [inline pkg/packagekit/updates.jsx:372]: For consistency, why does Update History use a caption, but this one uses a separate `<h3>`? I suppose we had our reasons back then, but maybe with the new tables this would work more consistently now?

Comment (garrett) [reply pkg/packagekit/updates.jsx:372]: Captions shouldn't be used. WebKit cannot render them properly and there are also some accessibility issues.

File: pkg/packagekit/history.jsx (hunk 11 above)
Comment (garrett) [reply pkg/packagekit/history.jsx:109]: Why are we using caption here? Does it render to `<caption>`? If so, we need to switch to using a heading. If "caption" actually gets rendered out to a heading, like `<h3>`, then that's fine, but it should be consistent in `updates.jsx` too, as @martinpitt points out below.

**Comment by garrett** [comment]
Updates isn't ready for the list. As per the discussion @KKoukiou and I had yesterday, it should be a follow-up PR, not in this one.

File: pkg/lib/cockpit-components-table.jsx [hunk 13]
```diff
@@ -56,8 +61,25 @@ export class ListingTable extends React.Component {
...
         }
-        this.state = { sortBy };
         this.onSort = this.onSort.bind(this);
+        this.onCollapse = this.onCollapse.bind(this);
+        this.reformatRows = this.reformatRows.bind(this);
+
+        this.state = { sortBy, isOpen: {} };
+    }
+
+    static getDerivedStateFromProps(nextProps, prevState) {
```
[the remark this answers is not in the document]
Comment (KKoukiou) [reply pkg/lib/cockpit-components-table.jsx:71]: The constructor gets run only the first time the component is accessed. Sometimes (noticed in the VMs page) this component gets initialized without any rows (empty table), and then after some seconds that the data are fetched from the API, the component is re-called with some rows. If I did not derive state from props in this case, we would skip updating the 'isOpen' state when new rows appear, after the component is mounted.

File: pkg/lib/listing.scss [hunk 14]
```diff
@@ -651,3 +664,8 @@ tbody tr.listing-ct-noexpand {
     /* 42 = (20px padding + 1px border) * 2 */
     max-width: calc(100vw - 42px);
 }
+
+// FIXME: (Garrett) table-grid.css rules mess with our form-layout.less
```
[the remark this answers is not in the document]
Comment (KKoukiou) [reply pkg/lib/listing.scss]: This was removed in lateron commit by @garrett , I rearranged the order so this is gone now.

File: pkg/machines/helpers.js [hunk 15]
```diff
@@ -24,6 +24,10 @@ const _ = cockpit.gettext;
 export const LIBVIRT_SESSION_CONNECTION = 'session';
 export const LIBVIRT_SYSTEM_CONNECTION = 'system';
 
+export function dummyVmsConvert(vms, uiVms) {
```
[the remark this answers is not in the document]
Comment (KKoukiou) [reply pkg/machines/helpers.js:27]: Indeed. But that function is not new - it was just moved out of dummyVm component.
I will replace the name anyway - it's confusing indeed.

File: pkg/machines/hostvmslist.jsx [hunk 16]
```diff
@@ -34,11 +34,27 @@ import {
...
+import { VmExpandedContent } from './components/vm/vmExpandedContent.jsx';
+import StateIcon from './components/vm/stateIcon.jsx';
+
+const VmState = ({ vm, resourceHasError }) => {
+    let state = null;
+
+    if (vm.installInProgress) {
+        state = 'creating VM installation';
+    } else if (vm.createInProgress) {
+        state = 'creating VM';
```
[the remark this answers is not in the document]
Comment (KKoukiou) [reply pkg/machines/hostvmslist.jsx:50]: These are humanly visible.

File: pkg/lib/listing.scss [hunk 17]
```diff
@@ -106,7 +106,6 @@ tbody > tr.listing-ct-item.listing-ct-warning {
     background-color: $listing-ct-warning-color;
 }
 
-table.pf-c-table tbody.pf-m-expanded > tr,
```
[the remark this answers is not in the document]
Comment (KKoukiou) [reply pkg/lib/listing.scss]: Fixed.

File: pkg/lib/cockpit-components-table.jsx (hunk 10 above)
Comment (KKoukiou) [reply pkg/lib/cockpit-components-table.jsx]: Basically if you have it in the default props it would be overwritten when the user will pass another className prop. Thus we need it hardcoded.
I addressed the comment with the props naming in a separate commit, since it's like that before this PR unfortunately

File: pkg/lib/cockpit-components-listing-panel.scss (hunk 9 above)
Comment (KKoukiou) [reply pkg/lib/cockpit-components-listing-panel.scss]: not present anymore.

File: pkg/machines/hostvmslist.jsx (hunk 16 above)
[the remark this answers is not in the document]
Comment (martinpitt) [reply pkg/machines/hostvmslist.jsx:50]: Then they need to be translated.

File: pkg/machines/components/vm/dummyVm.jsx [hunk 18]
```diff
@@ -1,64 +0,0 @@
...
-import { ListingRow } from "cockpit-components-listing.jsx";
-
-import {
-    rephraseUI,
-    vmId,
-} from "../../helpers.js";
-
-import StateIcon from './stateIcon.jsx';
-
-/** One Ui Dummy VM in the list (a row)
```
[the remark this answers is not in the document]
Comment (KKoukiou) [reply pkg/machines/components/vm/dummyVm.jsx:31]: So this is there forever. Someone decided to not have these unified although the fields we display are the same. (name, connection, state)
It's unnecessary complexity at least in the current state of the code.
Maybe when it was written it made more sense.

**Comment by KKoukiou** [comment]
@martinpitt this is clean from failures (had some distropkg ones which I addressed)
I also addressed your previous review.
I squashed all garretts commits into my major change - because they are interconnected - which created a bigger single commit but it's hard to split it now.

Would you like to continue with reviewing this one? Or shall I ask someone else?

File: pkg/lib/cockpit-components-listing-panel.scss [hunk 19]
```diff
@@ -0,0 +1,127 @@
+.ct-listing-panel {
```
Comment (marusak) [inline pkg/lib/cockpit-components-listing-panel.scss:1]: Admittedly I am not sure why you renamed `listing-ct-*` to `ct-listing-*`.
Just curious, was there some reason for that? (of course even if there was none, no reason to change it back :) )

File: pkg/machines/hostvmslist.jsx (hunk 16 above)
[the remark this answers is not in the document]
Comment (marusak) [reply pkg/machines/hostvmslist.jsx:50]: This was resolved, but they are still not translated.

File: pkg/lib/cockpit-components-listing-panel.jsx [hunk 20]
```diff
@@ -73,19 +71,17 @@ export class ListingPanel extends React.Component {
                 delete loadedTabs[prevTab];
 
             // ensure the new tab is loaded and update state
-            loadedTabs[tabIdx] = true;
-            this.setState({ loadedTabs: loadedTabs, activeTab: tabIdx });
+            loadedTabs[result.itemId] = true;
+            this.setState({ loadedTabs: loadedTabs, activeTab: result.itemId });
         }
-        e.stopPropagation();
-        e.preventDefault();
```
[the remark this answers is not in the document]
Comment (marusak) [reply pkg/lib/cockpit-components-listing-panel.jsx]: https://www.patternfly.org/v4/documentation/react/components/nav

NavItem has `preventDefault` property, which by default is false. Maybe it should be used in this case?

File: pkg/lib/cockpit-components-listing-panel.jsx [hunk 21]
```diff
@@ -73,19 +71,17 @@ export class ListingPanel extends React.Component {
...
 
     render() {
         const links = this.props.tabRenderers.map((itm, idx) => {
             return (
-                <li key={idx} className={ (idx === this.state.activeTab) ? "active" : ""}>
-                    <a href="#" tabIndex="0" onClick={ this.handleTabClick.bind(self, idx) }>{itm.name}</a>
-                </li>
+                <NavItem key={idx} id={itm.id} itemId={idx} isActive={idx === this.state.activeTab}>
+                    {itm.name}
+                </NavItem>
```
Comment (marusak) [inline pkg/lib/cockpit-components-listing-panel.jsx:85]: This is a11y regression. Now the navigation is not reachable by keyboard but it used to be.
I cannot make sense of this, as this NavItem is rendered as `div` but according to documentation it should be `a` by default (which would be accessible). Maybe it is worth reporting (or diving deep how it actually should work?)

We have similar pattern in selinux and it works there as expected.

File: pkg/lib/cockpit-components-listing-panel.scss (hunk 19 above)
Comment (KKoukiou) [reply pkg/lib/cockpit-components-listing-panel.scss:1]: This change was from @garrett, afaik he wants to keep all cockpit classes with ct-* prefix from now on.

**Comment by martinpitt** [comment]
There are still several Machines related [failures](https://logs.cockpit-project.org/logs/pull-13474-20200518-205318-e54c8efa-debian-testing/log.html).

**Comment by martinpitt** [comment]
My [earlier review](https://github.com/cockpit-project/cockpit/pull/13474#pullrequestreview-406336443) still has some open questions/points. Note that they are unhelpfully hidden by default, you need to expand them. The only relevant one is the change from `<span>` to an inaccessible `<a>`.

The "Network Interfaces" tab still looks wrong on more narrow screens, and causes horizontal scrolling:

image

and when you scroll right there, or make the window a little wider, the buttons are broken:

image

Only for very wide windows (wider than about 1300 px) the buttons look ok. In mobile mode, the table looks reasonable now (great work!)

**Comment by KKoukiou** [comment]
> My [earlier review](https://github.com/cockpit-project/cockpit/pull/13474#pullrequestreview-406336443) still has some open questions/points. Note that they are unhelpfully hidden by default, you need to expand them. The only relevant one is the change from `<span>` to an inaccessible `<a>`.
> 
> The "Network Interfaces" tab still looks wrong on more narrow screens, and causes horizontal scrolling:
> 
> image
> 
> and when you scroll right there, or make the window a little wider, the buttons are broken:
> 
> image
> 
> Only for very wide windows (wider than about 1300 px) the buttons look ok. In mobile mode, the table looks reasonable now (great work!)

@garrett this is because of
```
commit 64cefab29ccbd82142a6ea3b75f444aca24f220d
Author: Garrett LeSage <garrett@lesage.us>
Date:   Tue Apr 14 18:05:37 2020 +0200

    machines: Nested list buttons should wrap when needed

```

Can you please address it?

**Comment by garrett** [comment]
I've pushed my pending changes, with a bunch of them squashed.

It's still not ready. 

I really don't know how much more needs to be done, aside from the lists within lists now though... and I can't really tell until that's addressed, really. Should the content overflow? Probably. But it isn't doing so correctly, so data is gets truncated and there's no obvious way to get to it in some cases right now, unless we fix the sublists. Things are _definitely_ still broken at "smaller sizes", including medium to pretty large desktop sizes. Mobile is very messed up right now still, due to PF4 breaking everything apart. But, again, I think this is all related to the sublists.

I did address a bunch of issues, including the button overflow problem in the previous screenshot.

Meanwhile, switching tabs jumps back up to the top of the page, for some odd reason. Are they links without `preventDefault`?

**Comment by garrett** [comment]
In Firefox, switching tabs works. However, in Chrome, switching tabs to disks (specifcally) causes _all of_ the content to overflow and the page gets a scrollbar. I've attached screenshots which show this issue:

Overview | Disks (w/ overflow)
--- | ---
vm-overview | vm-disks

File: src/bridge/cockpitpipechannel.c [hunk 22]
```diff
@@ -415,11 +415,17 @@ cockpit_pipe_channel_prepare (CockpitChannel *channel)
...
+  gsize length;
+
   gint fd;
 
   COCKPIT_CHANNEL_CLASS (cockpit_pipe_channel_parent_class)->prepare (channel);
 
   options = cockpit_channel_get_options (channel);
+￼
+  debug_msg = cockpit_json_write_object (options, &length);
+  g_warning ("Debug Options: %s", debug_msg);
```
Comment (martinpitt) [inline src/bridge/cockpitpipechannel.c:428]: A warning will break every unit test that touches this, so maybe reduce it to a `g_message()`. But actually, I think it makes more sense to move this into the `cockpit_channel_fail()` call, like so:
```
        cockpit_channel_fail (channel, "protocol-error",
                            "invalid \"spawn\" option for stream channel: %s", cockpit_json_write_object (options, NULL));
```

**Comment by martinpitt** [comment]
I fixed the debugging patch (only build tested though), triggering tests to get a first set of results.

**Comment by martinpitt** [comment]
@KKoukiou : Hmm, something else happened here now -- a lot of Machines tests now fail, and that's completely unrelated to that spawn warning. Actually, none of the logs (and I looked at most) even contain that warning now.

I can perfectly reproduce the regressions locally, directly from your branch checkout. So the problem is not that the rebasing in CI causes this. The problem is rather that `tbody.open` does not exist at all any more, even though the [screenshot](https://logs.cockpit-project.org/logs/pull-13474-20200609-174502-7315f6f5-rhel-8-3/TestMachines-testCreateThenInstall-rhel-8-3-127.0.0.2-2201-FAIL.png) clearly shows it as open. That might be `tbody.pf-m-expanded` now?

**Comment by martinpitt** [comment]
5 out of the 6 failures were really simple, I pushed a fixup. Just [testConfigureBeforeInstall](https://logs.cockpit-project.org/logs/pull-13474-20200609-174502-7315f6f5-debian-stable/log.html#19-2) is weird -- locally it passes, but it consistently fails on CI, it seems the VM is unbootable, but it expects a second libvirt message in the log. I'll retry with rebasing to current master.

**Comment by martinpitt** [comment]
ah, I [caught the spawn failure](https://logs.cockpit-project.org/logs/pull-13474-20200609-195107-63d0c620-rhel-8-3/log.html#24)):
```
invalid "spawn" option for stream channel: {"payload":"stream","spawn":false,"pty":true,"superuser":"try","command":"open","channel":"2:5!19","flow-control":true,"group":"cockpit1:localhost/machines"}
```
So `false` is clearly bogus, but it's still not clear where this was sent from.

As expected, [testConfigureBeforeInstall](https://logs.cockpit-project.org/logs/pull-13474-20200609-195107-63d0c620-fedora-32/log.html#19-2) keeps failing. This reproduces perfectly well locally after rebasing.

**Comment by martinpitt** [comment]
I went through all usages of cockpit.spawn() in pkg/machines, and pkg/machines/components/consoles.jsx is the only candidate where this could happen:

```
const serialConsoleCommand = LibvirtDBus.serialConsoleCommand({ vm });

    return <SerialConsole connectionName={vm.connectionName} vmName={vm.name} spawnArgs={serialConsoleCommand}>{consoleSelector}</SerialConsole>;
```

Maybe this is related to 

    > warning: error: failed to get emulator capabilities

and thus libvirt serialConsoleCommand returns `false` ?

I pushed a possible fix.

**Comment by martinpitt** [comment]
Ah, and the tests need to be adjusted or skipped for [rhel-8-3-distrokpg](https://logs.cockpit-project.org/logs/pull-13474-20200609-211726-c276698f-rhel-8-3-distropkg/log.html)
